From 2d4d3d2b67fd5761c24b1fd2adc8707e5372b016 Mon Sep 17 00:00:00 2001
From: Jan Kiszka <jan.kiszka@siemens.com>
Date: Mon, 28 Feb 2022 17:06:20 +0100
Subject: [PATCH 17/23] binman: Include also subnodes in generator nodes

This allows to prefill fdt and config nodes with hash and signature
subnodes. It's just important to place the child nodes last so that
hashes do not come before the data - would be disliked by mkimage.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
Reviewed-by: Simon Glass <sjg@chromium.org>
[Jan: backported to v2022.01]
---
 tools/binman/etype/fit.py         | 3 +++
 tools/binman/ftest.py             | 1 +
 tools/binman/test/170_fit_fdt.dts | 3 +++
 3 files changed, 7 insertions(+)

diff --git a/tools/binman/etype/fit.py b/tools/binman/etype/fit.py
index 6936f5736a6..a7c6b870bd8 100644
--- a/tools/binman/etype/fit.py
+++ b/tools/binman/etype/fit.py
@@ -212,6 +212,9 @@ class Entry_fit(Entry):
                                 if depth == 1 and in_images:
                                     fsw.property('data',
                                                  tools.ReadFile(fname))
+                                for subsubnode in subnode.subnodes:
+                                    with fsw.add_node(subsubnode.name):
+                                        _AddNode(subnode, depth + 1, subsubnode)
                     else:
                         if self._fdts is None:
                             if self._fit_list_prop:
diff --git a/tools/binman/ftest.py b/tools/binman/ftest.py
index 2a98d810288..194ab0dca1b 100644
--- a/tools/binman/ftest.py
+++ b/tools/binman/ftest.py
@@ -3823,6 +3823,7 @@ class TestFunctional(unittest.TestCase):
             self.assertEqual(expected_data, fnode.props['data'].bytes)
             self.assertEqual('fdt-test-fdt%d.dtb' % seq,
                              fnode.props['description'].value)
+            self.assertEqual(fnode.subnodes[0].name, 'hash')
 
         def _CheckConfig(seq, expected_data):
             """Check the configuration nodes
diff --git a/tools/binman/test/170_fit_fdt.dts b/tools/binman/test/170_fit_fdt.dts
index 99d710c57e9..0197ffd1597 100644
--- a/tools/binman/test/170_fit_fdt.dts
+++ b/tools/binman/test/170_fit_fdt.dts
@@ -36,6 +36,9 @@
 					description = "fdt-NAME.dtb";
 					type = "flat_dt";
 					compression = "none";
+					hash {
+						algo = "sha256";
+					};
 				};
 			};
 
-- 
2.34.1

